grpc/src/test/java/com/linecorp/armeria/server/grpc/GrpcServiceImplErrorTest.java

```diff
@@ -0,0 +1,319 @@
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.google.protobuf.ByteString;
+
+import com.linecorp.armeria.client.ClientRequestContextCaptor;
+import com.linecorp.armeria.client.Clients;
+import com.linecorp.armeria.client.grpc.GrpcClients;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.grpc.protocol.GrpcHeaderNames;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.grpc.testing.Messages.Payload;
+import com.linecorp.armeria.grpc.testing.Messages.SimpleRequest;
+import com.linecorp.armeria.grpc.testing.Messages.SimpleResponse;
+import com.linecorp.armeria.grpc.testing.Messages.StreamingOutputCallRequest;
+import com.linecorp.armeria.grpc.testing.Messages.StreamingOutputCallResponse;
+import com.linecorp.armeria.grpc.testing.TestServiceGrpc;
+import com.linecorp.armeria.grpc.testing.TestServiceGrpc.TestServiceBlockingStub;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.logging.LoggingService;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.grpc.ForwardingServerCall.SimpleForwardingServerCall;
+import io.grpc.InternalMetadata;
+import io.grpc.Metadata;
+import io.grpc.Metadata.Key;
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+
+class GrpcServiceImplErrorTest {
+
+    private static final StreamingOutputCallRequest STREAMING_OUTPUT_CALL_REQUEST_MESSAGE =
+            StreamingOutputCallRequest
+                    .newBuilder()
+                    .setPayload(
+                            Payload.newBuilder().setBody(ByteString.copyFromUtf8("ping1")).build())
+                    .build();
+    // Valid metadata has even count of binaryValues.
+    private static final Metadata validMetadata = InternalMetadata.newMetadata(
+            "key1".getBytes(StandardCharsets.US_ASCII),
+            "val1".getBytes(StandardCharsets.US_ASCII),
+            "key2".getBytes(StandardCharsets.US_ASCII),
+            "val2".getBytes(StandardCharsets.US_ASCII)
+    );
+    // 'usedNames' is 3, but size of 'binaryValues' is 2.
+    // 'usedNames' and size of 'binaryValues' must be equal, so this is corrupted metadata.
+    private static final Metadata corruptedMetadata =
+            InternalMetadata.newMetadata(
+                    3, "key1".getBytes(StandardCharsets.US_ASCII), "val1".getBytes(StandardCharsets.US_ASCII));
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            configureGrpcService(sb);
+        }
+    };
+
+    @RegisterExtension
+    static ServerExtension serverUsingCorruptedInterceptor = new ServerExtension() {
```

Comment:
We may use a single `server` by doing the following:

1. modify `configureGrpcService` to accept an extra prefix
```
private static void configureGrpcService(
    String prefix, ServerBuilder sb, ServerInterceptor... interceptors) {
sb.serviceUnder(prefix, GrpcService.builder().addService(new ServerImplErrorAtMetadataService())
                      .intercept(interceptors).build());
sb.decorator(LoggingService.newDecorator());
}
```

2. Add a service with the interceptor:
```
configureGrpcService("/interceptor", sb, corruptedServerInterceptor);
```

3. Call with a prefix from clients if needed
```
final TestServiceBlockingStub client =
        GrpcClients.builder("http://127.0.0.1:" + server.httpPort())
                   .pathPrefix("/interceptor")
                   .build(TestServiceBlockingStub.class);
```

Reply:
This looks good, because this follows Armeria's `decorator` visitor design pattern.
I'll try to apply your suggestion.